src/iceberg/schema.cc

```diff
@@ -257,4 +258,268 @@ void NameToIdVisitor::Finish() {
   }
 }
 
+/// \brief Visitor class for pruning schema columns based on selected field IDs.
+///
+/// This visitor traverses a schema and creates a projected version containing only
+/// the specified fields. It handles different projection modes:
+/// - select_full_types=true: Include entire fields when their ID is selected
+/// - select_full_types=false: Recursively project nested fields within selected structs
+///
+/// \warning Error conditions that will cause projection to fail:
+/// - Attempting to explicitly project List or Map types (returns InvalidArgument)
+/// - Projecting a List when element result is null (returns InvalidArgument)
+/// - Projecting a Map without a defined map value type (returns InvalidArgument)
+/// - Projecting a struct when result is not StructType (returns InvalidArgument)
+class PruneColumnVisitor {
+ public:
+  PruneColumnVisitor(const std::unordered_set<int32_t>& selected_ids,
+                     bool select_full_types)
+      : selected_ids_(selected_ids), select_full_types_(select_full_types) {}
+
+  Status Visit(const StructType& type, std::unique_ptr<Type>* out) const {
```

Comment:
```suggestion
  Status Visit(const std::shared_ptr<StructType>& type, std::shared_ptr<Type>* out) const {
```

I still think we need to change the signature to this to enable returning original type (which is a common case for select). Or even better, we can simplify it as `Result<std::shared_ptr<Type>> Visit(const std::shared_ptr<StructType>& type) const` and add our own dispatching in the `Result<std::shared_ptr<Type>> Visit(const std::shared_ptr<Type>& type) const`.